presto-spark-base/src/main/java/com/facebook/presto/spark/PrestoSparkQueryExecutionFactory.java

```diff
@@ -385,29 +403,54 @@ private PrestoSparkQueryExecution(
                 SerializedPrestoSparkTaskDescriptor serializedTaskDescriptor = new SerializedPrestoSparkTaskDescriptor(sparkTaskDescriptorJsonCodec.toJsonBytes(taskDescriptor));
 
                 SubPlan child = getOnlyElement(root.getChildren());
-                JavaPairRDD<Integer, PrestoSparkRow> rdd = createRdd(child);
+                JavaPairRDD<Integer, PrestoSparkRow> rdd = createRdd(child, broadcastsToDestroy);
                 List<Tuple2<Integer, PrestoSparkRow>> sparkDriverInput = rdd.collect();
+                broadcastsToDestroy.forEach(Broadcast::destroy);
                 return ImmutableList.copyOf(executorFactoryProvider.get().create(
                         0,
                         0,
                         serializedTaskDescriptor,
-                        new PrestoSparkTaskInputs(ImmutableMap.of(child.getFragment().getId().toString(), sparkDriverInput.iterator())),
+                        new PrestoSparkTaskInputs(ImmutableMap.of(child.getFragment().getId().toString(), sparkDriverInput.iterator()), ImmutableMap.of()),
                         taskStatsCollector));
             }
 
-            return createRdd(root).collect();
+            List<Tuple2<Integer, PrestoSparkRow>> result = createRdd(root, broadcastsToDestroy).collect();
+            broadcastsToDestroy.forEach(Broadcast::destroy);
+            return result;
         }
 
-        private JavaPairRDD<Integer, PrestoSparkRow> createRdd(SubPlan subPlan)
+        private JavaPairRDD<Integer, PrestoSparkRow> createRdd(SubPlan subPlan, List<Broadcast<?>> broadcasts)
```

Comment:
It looks to me this method actually returns two values:
- An RDD `JavaPairRDD<Integer, PrestoSparkRow>` that transform Presto `SubPlan` to Spark RDD.
- A List of `Broadcast`, indicates "we have done these Broadcast in preparing for executing this RDD". As a result,  once the returned RDD finish execution, the broadcasts should be destroyed. 

However, this seems a bit difficult to understand -- as the `broadcasts` list is stateful and get appended in the recursive call. I am wondering if it is possible to have a `RddAndBroadcast` class to return both the RDD and the broadcasts?